device.go

```diff
@@ -158,11 +166,55 @@ func newBlockDevice(devInfo DeviceInfo) *BlockDevice {
 	}
 }
 
-func (device *BlockDevice) attach(h hypervisor) error {
+func (device *BlockDevice) attach(h hypervisor, c *Container, create bool) error {
+	device.DeviceInfo.ID = uuid.Generate().String()
+	devID := fmt.Sprintf("drive-%s", device.DeviceInfo.ID)
+
+	drive := Drive{
+		File:   device.DeviceInfo.HostPath,
+		Format: "raw",
+		ID:     devID,
+	}
+
+	// Increment the block index for the pod. This is used to determine the name
+	// for the block device in the case where the block device is used as container
+	// rootfs and the predicted block device name needs to be provided to the agent.
+	_, err := c.pod.getAndSetPodBlockIndex()
+	if err != nil {
+		return err
+	}
+
+	virtLog.Infof("Attaching block device %s", device.DeviceInfo.HostPath)
+	if create {
+		if err := h.addDevice(drive, blockDev); err != nil {
+			return err
+		}
+		device.DeviceInfo.Hotplugged = false
+	} else {
+		if err := h.hotplugAddDevice(drive, blockDev); err != nil {
+			return err
+		}
+		device.DeviceInfo.Hotplugged = true
+	}
+
 	return nil
 }
 
 func (device BlockDevice) detach(h hypervisor) error {
+	if device.DeviceInfo.Hotplugged {
+		virtLog.Infof("Unplugging block device %s", device.DeviceInfo.HostPath)
+
+		devID := fmt.Sprintf("drive-%s", device.DeviceInfo.ID)
+		drive := Drive{
+			ID: devID,
+		}
+
+		if err := h.hotplugRemoveDevice(drive, blockDev); err != nil {
```

Comment:
I think the block device can be cold plugged, why hot plug it?

Reply:
You need to be able to hotplug it in case you are adding a block device through a new container in case of an existing pod. The VM will already be up and running, and you will need to hotplug this new block device related to this new container. Makes sense ?

Reply:
@sboeuf yes
one more question, I not able to see the block device inside the container, I guess there is something missing in the agent code (re-mount /dev/?)

Reply:
Now that I am passing the mount information to the agent in a following commit, this will be resolved, as the agent will handle mounting the devices passed.